File: 剑指offer_java/剑指 Offer 21. 调整数组顺序使奇数位于偶数前面.java
class Solution {
    public int[] exchange(int[] nums) {
        int i = 0, j = nums.length - 1, tmp;
        while(i < j) {
            while(i < j && (nums[i] & 1) == 1) i++;
            while(i < j && (nums[j] & 1) == 0) j--;
            tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
        }
        return nums;
    }
}

class Solution {
    public int[] exchange(int[] nums) {
        int slow = 0, fast = 0;
        while(fast < nums.length){
            if(nums[fast] % 2 == 1){
                int tmp = nums[slow];
                nums[slow] = nums[fast];
                nums[fast] = tmp;
                slow++;
            }
            fast++;
        }
        return nums;
    }
}

File: 剑指offer/剑指 Offer 21. 调整数组顺序使奇数位于偶数前面.cpp

class Solution {
public:
    vector<int> exchange(vector<int>& nums) {
        int left =0,right=nums.size()-1;
        while(left<right){
            if((nums[left]&1)!=0){
                left++;
                continue;
            }
            if((nums[right]&1)!=1){
                right--;
                continue;
            }
            swap(nums[left++],nums[right--]);
          }
          return nums;
    }
};


//快慢指针 fast找奇数 slow是偶数

class Solution {
public:
    vector<int> exchange(vector<int>& nums) {
        int slow=0, fast=0;
        while(fast<nums.size()){
            if(nums[fast]&1){
            swap(nums[slow],nums[fast]);
            slow++;
            }
            fast ++;
        }
return nums;
    }

};
